elastic/kibana
# [Fleet] Use multiple saved object type for agent and package policies with opt-in migration

## Description 

Related to https://github.com/elastic/kibana/issues/188420

Add an opt-in migration to enable space awareness, and use new saved object type to get multiple space affinity for agent and package policy.


## Details

### Storing feature state

The opt-in is done via the fleet settings saved object `use_space_awareness_migration_status`.

I introduced a new helper `isSpaceAwarenessEnabled` to know if a user opt-in for space awareness that method return true if the feature flag is set and the migration is complete. 


### Saved object

I introduced a new saved object type `.fleet-agent-policies` that is space aware (`multiple`). The service that query that saved object use either the space aware type or `ingest-agent-policies` based on the settings state.

I did the same for the package policies.

I also use the dynamic saved object type in the diverse method that directly access the saved objects.

The kuery is rewritten to match the used object type so both the old and the new saved object type are now working.


### Migration

 I introduced a new API endpoint `/internal/fleet/enable_space_awareness`

That endpoint set `use_space_awareness_status` and copy the agent policies from `.ingest-agent-policies` to `.fleet-agent-policies` same for package policies.

The API acces to `/api/fleet` is locked during the migration as the package policy write methods, to avoid user loosing data.


### How to test

with the feature flag `useSpaceAwareness` enabled.

You can enable space awareness 
```
 curl -u elastic:changeme -XPOST "http://localhost:5601/internal/fleet/enable_space_awareness" -H "kbn-xsrf: reporting" -H 'elastic-api-version: 1'
 ```
 
Than verify that policies created in different space are not available in others.

You can also check there is no issue when not enabling space awareness.


### Test sharing 

You can test sharing a saved object between multiple saved by updating the namespaces field of the saved object and check it's visible in both space, it seems to work as expected.

```
"namespaces": [
            "default",
            "test"
          ],
```


## Todo

- [x] Replace use of feature flag, by `isSpaceAwareness` helper
- [x] Lock package policy
- [x] API integration
  - [x] enable space awareness for all existing space tests
  - [x] Test migration

## Not done in that PR 

- UI to opt-in for space awareness
- Migrating uninstall tokens and message signin saved object, as we probably are going to not use a space saved object type for them.
- Surface the support for multiple space in API/UI, this will be done in https://github.com/elastic/kibana/pull/189663
